https://nemu-jenkins-ci.westus2.cloudapp.azure.com/job/nemu/job/PR-141/6/consoleFull

```
[xenial-virt] --- FAIL: TestNemu (0.00s)
[xenial-virt]     --- FAIL: TestNemu/PCIHotplug/virt/clear (28.29s)
[xenial-virt]     	nemu_test.go:178: launching /home/agentadmin/build-x86_64_virt/x86_64_virt-softmmu/qemu-system-x86_64_virt with: [-machine virt,accel=kvm,kernel_irqchip,nvdimm -bios /home/agentadmin/workloads/OVMF.fd -smp 2,cores=1,threads=1,sockets=2,maxcpus=32 -m 512,slots=4,maxmem=16384M -cpu host -nographic -no-user-config -nodefaults -drive file=/tmp/nemu-test225261334,if=none,id=drive-virtio-disk0,format=qcow2 -device virtio-blk-pci,scsi=off,drive=drive-virtio-disk0,id=virtio-disk0 -device virtio-blk-pci,drive=cloud -drive if=none,id=cloud,file=/tmp/nemu-test243394924,format=raw -netdev user,id=mynet0,hostfwd=tcp::2224-:22,hostname=nemuvm -device virtio-net-pci,netdev=mynet0 -device virtio-serial-pci,id=virtio-serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev file,id=charconsole0,path=/tmp/nemu-test219218899,server,nowait -device virtio-rng-pci,rng=rng0 -object rng-random,filename=/dev/random,id=rng0 -device virtio-balloon-pci -object cryptodev-backend-builtin,id=cryptodev0 -device virtio-crypto-pci,id=crypto0,cryptodev=cryptodev0 -device sysbus-debugcon,iobase=0x402,chardev=debugcon -chardev file,path=/tmp/nemu-test737456953,id=debugcon -device sysbus-debugcon,iobase=0x3f8,chardev=serialcon -chardev file,path=/tmp/nemu-test219032388,id=serialcon -qmp unix:/tmp/nemu-test758400864,server,nowait]
[xenial-virt]     	nemu_test.go:218: 
[xenial-virt]     		QMP version: {3 0 0 []}
[xenial-virt]     	nemu_test.go:178: {"execute":"qmp_capabilities"}
[xenial-virt]     	nemu_test.go:178: {"arguments":{"addr":"12","driver":"virtio-net-pci","id":"net1","mac":"ab:cd:ef:01:23:45","romfile":""},"execute":"device_add"}
[xenial-virt]     	nemu_test.go:163: SSH command generated error: Process exited with status 1
[xenial-virt]     	nemu_amd64_test.go:241: PCI device not detected from the guest
[xenial-virt]     	nemu_test.go:178: {"execute":"quit"}
```

--- a/tools/CI/nats/nemu_amd64_test.go
+++ b/tools/CI/nats/nemu_amd64_test.go
@@ -223,6 +223,44 @@ func testCPUHotplug(ctx context.Context, q *qemuTest, t *testing.T) {
 	}
 }
 
+// testPCIHotplug can only be tested for pc and virt machine types since q35
+// needs a PCIe hierarchy, meaning we need to add a root port to the initial
+// VM.
+func testPCIHotplug(ctx context.Context, q *qemuTest, t *testing.T) {
+	// Use virtio-net-pci device for PCI hotplug.
+	addr := "12"
+	macAddr := "ab:cd:ef:01:23:45"
+	err := q.qmp.ExecuteNetPCIDeviceAdd(ctx, "", "net1", macAddr, addr, "", "", 0)
+	if err != nil {
+		t.Errorf("Error hotplugging PCI device (virtio-net-pci): %v", err)
+	}
+
+	// This sleep ensures the guest OS has enough time to detect the new
+	// PCI device and create the network interface.
+	time.Sleep(time.Second * 30)
+
+	// Check the PCI device has been detected by the guest OS.
+	expected := "exist"
+	cmd := fmt.Sprintf("ls /sys/bus/pci/devices/0000:00:%s.0 &> /dev/null && echo %s", addr, expected)
+	result := strings.TrimSpace(q.runCommandBySSH(cmd, t))
+	if result != expected {
+		t.Error("PCI device not detected")
+	}
+
+	// Check the network interface has been created.
+	cmd = fmt.Sprintf("ip a | grep '%s' &> /dev/null && echo %s", macAddr, expected)
+	result = strings.TrimSpace(q.runCommandBySSH(cmd, t))
+	if result != expected {
+		t.Error("Network interface not created")
+	}
+
+	time.Sleep(time.Second * 15)
+	err = q.qmp.ExecuteQuit(ctx)
+	if err != nil {
+		t.Errorf("Error quiting via QMP: %v", err)
+	}
+}
+
 var machines = []string{"pc", "q35", "virt"}
 
 const (
@@ -297,5 +335,11 @@ var tests = []testConfig{
 		testFunc: testMemoryHotplug,
 		distros:  clearLinuxOnly,
 		machines: allMachines,
+	},
+	{
+		name:     "PCIHotplug",
+		testFunc: testPCIHotplug,
+		distros:  clearLinuxOnly,
+		machines: []string{"pc", "virt"},
 	},
 }

